src/inputfetcher.h

```diff
@@ -0,0 +1,237 @@
+                if (!m_db->GetCoin(outpoint, coin)) {
+                    // Missing an input, just break. This block will fail validation, so no point in continuing.
+                    break;
+                }
+                pairs.emplace_back(std::move(outpoint), std::move(coin));
+            }
+        } while (true);
+    }
+
+    //! Add a batch of outpoints to the queue
+    void Add(std::vector<COutPoint>&& outpoints) noexcept EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
+    {
+        if (outpoints.empty()) {
+            return;
+        }
+
+        const auto size{outpoints.size()};
+        {
+            LOCK(m_mutex);
+            m_in_flight_fetches_count += outpoints.size();
+            if (m_outpoints.empty()) {
+                m_outpoints = std::move(outpoints);
+            } else {
+                m_outpoints.insert(m_outpoints.end(), std::make_move_iterator(outpoints.begin()), std::make_move_iterator(outpoints.end()));
+            }
+        }
+
+        if (size == 1) {
+            m_worker_cv.notify_one();
+        } else {
+            m_worker_cv.notify_all();
+        }
+    }
+
+
+public:
+    //! Create a new input fetcher
+    explicit InputFetcher(size_t batch_size, size_t worker_thread_count) noexcept
+        : m_batch_size(batch_size)
+    {
+        m_worker_threads.reserve(worker_thread_count);
+        for (size_t n = 0; n < worker_thread_count; ++n) {
+            m_worker_threads.emplace_back([this, n]() {
+                util::ThreadRename(strprintf("inputfetch.%i", n));
+                Loop();
+            });
+        }
+    }
+
+    // Since this class manages its own resources, which is a thread
+    // pool `m_worker_threads`, copy and move operations are not appropriate.
+    InputFetcher(const InputFetcher&) = delete;
+    InputFetcher& operator=(const InputFetcher&) = delete;
+    InputFetcher(InputFetcher&&) = delete;
+    InputFetcher& operator=(InputFetcher&&) = delete;
+
+
+    //! Fetch all block inputs from db, and insert into cache.
+    void FetchInputs(CCoinsViewCache& cache, const CCoinsViewDB& db, const CBlock& block) noexcept EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
+    {
+        m_db = &db;
+
+        std::vector<COutPoint> buffer{};
+        buffer.reserve(m_batch_size);
+        std::set<Txid> txids{};
+        for (const auto& tx : block.vtx) {
+            if (tx->IsCoinBase()) continue;
+            for (const auto& in : tx->vin) {
+                const auto& outpoint = in.prevout;
+                // If an input references an outpoint from earlier in the
+                // block, it won't be in the cache yet but it also won't be
+                // in the db either.
+                if (txids.contains(outpoint.hash)) {
+                    continue;
+                }
+                if (cache.HaveCoinInCache(outpoint)) {
+                    continue;
+                }
+
+                buffer.emplace_back(outpoint);
+                if (buffer.size() == m_batch_size) {
```

Comment:
Would it be possible to create the batch sizes dynamically?
Since the number of missing values differs for every block (and every dbcache size), it may not make more sense to calculate the optimal split instead of using a random value of 128.
Coroutines might alleviate this problem.

Reply:
I'm not sure it would warrant the complexity I think this batch size is "good enough" for now. In a follow up we could maybe add ways to set this with configs to experiment if there really is more optimal settings.